contracts/v3/FiatTokenV3.sol

```diff
@@ -0,0 +1,276 @@
+ */
+contract FiatTokenV3 is FiatTokenV2_1 {
+    mapping(address => uint256) internal frozenBalances;
+
+    event BalanceFrozen(address indexed _account, uint256 amountFrozen);
+    event BalanceUnfrozen(address indexed _account, uint256 amountUnfrozen);
+
+    /**
+     * @notice Initialize v3
+     */
+    function initializeV3() external {
+        require(_initializedVersion == 2, "v3 initialized out of order");
+
+        DOMAIN_SEPARATOR = EIP712.makeDomainSeparator(name, "3");
+
+        _initializedVersion = 3;
+    }
+
+    /**
+     * @dev Freezes an account's balance
+     * @param _account The address whose balance will be frozen
+     */
+    function freezeBalance(address _account) external onlyBlacklister {
+        uint256 amountFrozen = balances[_account];
+
+        totalSupply_ = totalSupply_.sub(amountFrozen);
+        frozenBalances[_account] = frozenBalances[_account].add(amountFrozen);
+        balances[_account] = 0;
+
+        emit BalanceFrozen(_account, amountFrozen);
+        emit Transfer(_account, address(0), amountFrozen);
+    }
+
+    /**
+     * @dev Unfreezes an account's balance
+     * @param _account The address whose balance will be unfrozen
+     */
+    function unfreezeBalance(address _account) external onlyBlacklister {
+        uint256 amountUnfrozen = frozenBalances[_account];
+
+        totalSupply_ = totalSupply_.add(amountUnfrozen);
+        frozenBalances[_account] = 0;
+        balances[_account] = balances[_account].add(amountUnfrozen);
+
+        emit BalanceUnfrozen(_account, amountUnfrozen);
+        emit Transfer(address(0), _account, amountUnfrozen);
+    }
+
+    /**
+     * @dev Get frozen token balance of an account
+     * @param account address The account
+     */
+    function frozenBalanceOf(address account) external view returns (uint256) {
+        return frozenBalances[account];
+    }
+
+    /**
+     * @notice Set spender's allowance over the caller's tokens to be a given
+     * value.
+     * @param spender   Spender's address
+     * @param value     Allowance amount
+     * @return True if successful
+     */
+    function approve(address spender, uint256 value)
+        external
+        override
+        whenNotPaused
+        returns (bool)
+    {
+        _approve(msg.sender, spender, value);
+        return true;
+    }
+
+    /**
+     * @notice Transfer tokens by spending allowance
+     * @param from  Payer's address
+     * @param to    Payee's address
+     * @param value Transfer amount
+     * @return True if successful
+     */
+    function transferFrom(
+        address from,
+        address to,
+        uint256 value
+    ) external override whenNotPaused returns (bool) {
+        require(
+            value <= allowed[from][msg.sender],
+            "ERC20: transfer amount exceeds allowance"
+        );
+        _transfer(from, to, value);
+        allowed[from][msg.sender] = allowed[from][msg.sender].sub(value);
```

Comment:
While you're at it, can save 5k more gas on swaps and contract interactions by using infinite approval gas optimization!!

Something like this:

```suggestion
        if (allowed[from][msg.sender != uint(-1)) {
            allowed[from][msg.sender] = allowed[from][msg.sender].sub(value);
        }
```

Reference
 - https://twitter.com/philipliao_/status/1479860204577693700?s=20
 - https://github.com/makerdao/dss/blob/master/src/dai.sol#L76

Reply:
in this moment I am euphoric

Reply:
@phil-ociraptor in your suggestion aren't you missing a `]`?

L121 of your suggestion 👇
```
if (allowed[from][msg.sender] != uint(-1)) {
```

Reply:
🦅 👀 

great catch! updated